src/main/java/vg/civcraft/mc/citadel/activity/ActivityMap.java

```diff
@@ -0,0 +1,209 @@
+											.prepareStatement(GET_ACTIVITY);
+									statement.setInt(1, groupId);
+									statement.setShort(2, id);
+									statement.setInt(3, resolution);
+									ResultSet set = statement.executeQuery();
+									while (set.next()) {
+										int scaledX = set.getInt("x");
+										int scaledZ = set.getInt("z");
+										list.put(scaledX, scaledZ, set.getTimestamp("activity").toInstant());
+									}
+								}
+								return list;
+							}
+						})
+		);
+	}
+
+	public void update(World world, Iterable<Integer> groups, int scaledX, int scaledZ) {
+		var cache = getCache(world);
+		for (int i = -radius; i <= radius; i++) {
+			for (int j = -radius; j <= radius; j++) {
+				try {
+					for (int group : groups) {
+						Table<Integer, Integer, Instant> list = cache.get(group);
+						int nx = scaledX + i;
+						int nz = scaledZ + j;
+						Instant now = Instant.now();
+						synchronized (list) {
+							list.put(nx, nz, now);
+						}
+
+						this.updates.add(new Update(worldIdManager.getInternalWorldId(world),
+								group, nx, nz, now));
+					}
+				} catch (ExecutionException e) {
+					e.printStackTrace();
+				}
+			}
+		}
+	}
+
+	public Optional<Instant> getLastActivityTime(Group group, Location location) {
+		if (!isEnabled(location.getWorld())) {
+			return Optional.of(Instant.ofEpochMilli(group.getActivityTimeStamp()));
+		}
+
+		try {
+			Table<Integer, Integer, Instant> activities = getCache(location.getWorld()).get(group.getGroupId());
+
+			int scaledX = location.getBlockX() / resolution;
+			int scaledZ = location.getBlockZ() / resolution;
+
+			synchronized (activities) {
+				Instant get = activities.get(scaledX, scaledZ);
+				if (get == null) {
+					Instant activity = Instant.ofEpochMilli(group.getActivityTimeStamp());
+					if (activity.compareTo(defaultActivity) <= 0) {
+						// If the group was last active before activity map was added, use the group's activity time for decay
+						get = activity;
+					} else {
+						// If the group has been active since, use the activity map decay
+						get = defaultActivity;
+					}
+				}
+				return Optional.of(get);
+			}
+		} catch (ExecutionException e) {
+			e.printStackTrace();
+			return Optional.empty();
+		}
+	}
+
+	private static class Update {
+
+		private final short world;
+		private final int group;
+		private final int x;
+		private final int z;
+		private final Instant timestamp;
+
+		private Update(short world, int group, int x, int z, Instant timestamp) {
+			this.world = world;
+			this.group = group;
+			this.x = x;
+			this.z = z;
+			this.timestamp = timestamp;
+		}
+	}
+
+	/*private static class TimestampedBox extends QTBoxImpl {
+
+		private final Instant instant;
+
+		public TimestampedBox(Instant instant, int x, int z) {
+			super(x, z, x + 1, z + 1);
+			this.instant = instant;
+		}
+	}*/
+
+	public boolean isEnabled(World world) {
+		return world != null && this.worlds.contains(world.getName());
```

Comment:
Would be nicer to check whether the world exists on startup (give a nice error message if not) and then do UUID based lookup instead of string comparison every time.

Reply:
what

Reply:
Right now everytime someones crosses a decay boundary, you do a string based lookup. Doing one based on UUID would be more performant. Further it'd also be nice if you checked on startup whether the world listed in the config even exist.

Reply:
are you seriously complaining about the performance of a hashset.contains

Reply:
Are you seriously complaining about Max's suggestion?

Reply:
Optimization is actually a relatively large concern for us when we’re already pushing the boundaries of what Minecraft servers can handle